File: Sorting Algorithm/quickSort.c
// utility: swap
void swap(int* array, int L, int R) {
    int temp = array[L];
    array[L] = array[R];
    array[R] = temp;

    return;
}

// utility: partition
int partition(int* array, int low, int high) {
    //parameters
    int pivot = low;
    int i = low;
    int j = high;

    //iterate and swap untill indices cross
    while (i < j) {
        while ((array[i] <= array[pivot]) && (i < high))
            ++i;
        while (array[j] > array[pivot])
            --j;
        if (i < j)
            swap(array, i, j);
    }
    swap(array, pivot, j);

    return j;
}

// algorithm
void quick_sort(int* array, int low, int high) {
    if (low >= high)
        return;

    int pivot = partition(array, low, high);

    quick_sort(array, low, pivot - 1);
    quick_sort(array, pivot + 1, high);

    return;
}

File: Sorting Algorithm/QuickSort.js
// Author: Julio Cesar Goldschmidt
// GitHub: https://github.com/JulioGold

function swap(items, firstIndex, secondIndex){
    var temp = items[firstIndex];
    items[firstIndex] = items[secondIndex];
    items[secondIndex] = temp;
}

function partition(items, left, right) {
    let pivot = items[Math.floor((right + left) / 2)];
    let i = left;
    let j = right;

    while (i <= j) {
        while (items[i] < pivot) {
            i++;
        }

        while (items[j] > pivot) {
            j--;
        }

        if (i <= j) {
            swap(items, i, j);
            i++;
            j--;
        }
    }

    return i;
}

/**
 * The method sort array with the QuickSort sort algorithm.
 * @param {Array} array Array of elements to sort.
 */
function QuickSort(items, left, right) {

    var index;

    if (items.length > 1) {

        left = typeof left != "number" ? 0 : left;
        right = typeof right != "number" ? items.length - 1 : right;

        index = partition(items, left, right);

        if (left < index - 1) {
            QuickSort(items, left, index - 1);
        }

        if (index < right) {
            QuickSort(items, index, right);
        }

    }

    return items;
}

/**
 * The main function to run the algorithm.
 */
(function main() {
    let arr = [3, 1, 2, 5, 4];
    let arrResult = QuickSort(arr);
    console.log(arrResult.join(','));
})()
